umbral/umbral.py

```diff
@@ -115,98 +116,135 @@ def __bytes__(self):
+                "Can't make a Capsule from nothing.  Pass either Alice's data (ie, point_eph_e) or Bob's (e_prime). \
+                Passing both is also fine.")
+
+        self._point_eph_e = point_eph_e
+        self._point_eph_v = point_eph_v
+        self._bn_sig = bn_sig
+
+        self._point_eph_e_prime = e_prime
+        self._point_eph_v_prime = v_prime
+        self._point_noninteractive = noninteractive_point
 
         self.cfrags = {}
+        self._contents = None
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    @classmethod
+    def from_original_bytes(cls, data: bytes, curve: ec.EllipticCurve):
         """
         Instantiates a Capsule object from the serialized data.
         """
         eph_e = Point.from_bytes(data[0:33], curve)
         eph_v = Point.from_bytes(data[33:66], curve)
         sig = BigNum.from_bytes(data[66:98], curve)
 
-        return Capsule(eph_e, eph_v, sig)
+        return cls(eph_e, eph_v, sig)
+
+    @classmethod
+    def from_reconstructed_bytes(cls, data: bytes, curve: ec.EllipticCurve):
+        """
+        Instantiate a Capsule from serialized data after reconstruction has occurred.
+
+        The most obvious use case is Bob affixing at least m cFrags and then serializing the Capsule.
+        """
+        # TODO: Seems like a job for BytestringSplitter ?
+        e_prime = Point.from_bytes(data[0:33], curve)
+        v_prime = Point.from_bytes(data[33:66], curve)
+        eph_ni = Point.from_bytes(data[66:99], curve)
+
+        return cls(e_prime=e_prime, v_prime=v_prime, noninteractive_point=eph_ni)
+
+    @property
+    def contents(self):
+        return self._contents
 
     def to_bytes(self):
         """
         Serialize the Capsule into a bytestring.
         """
-        eph_e = self.point_eph_e.to_bytes()
-        eph_v = self.point_eph_v.to_bytes()
-        sig = self.bn_sig.to_bytes()
+        eph_e = self._point_eph_e.to_bytes()
+        eph_v = self._point_eph_v.to_bytes()
+        sig = self._bn_sig.to_bytes()
 
         return eph_e + eph_v + sig
 
     def verify(self, params: UmbralParameters):
 
-        e = self.point_eph_e
-        v = self.point_eph_v
-        s = self.bn_sig
+        e = self._point_eph_e
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return params.g * s == v + (e * h)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def original_ephemeral_points(self):
+        return self._point_eph_e + self._point_eph_v
+
+    def reconstructed_ephemeral_points(self):
+        return self._point_eph_e_prime + self._point_eph_v_prime
+
+    def open(self, pub_bob, priv_bob, pub_alice, force_reopen=False, pre=None):
+        # TODO: Raise an error here if Bob has gathered enough cFrags.
+        if self.contents and not force_reopen:
+            newly_opened = True
+        else:
```

Comment:
This `newly_opened` and `force_reopen` doesn't really do much for me. It just adds some weird logic in an otherwise plain reference implementation. I don't understand what problem it's solving.

Reply:
Yeah I mostly agree.  I just didn't want to leave the user without this functionality without at least sketching and discussing it first.

Reply:
I think this should probably stay out of this PR until David and I discuss how we're going to handle the symmetric operations when something like this might be possible.